os/arch/arm/src/armv7-m/up_assert.c

```diff
@@ -250,95 +251,94 @@ static void up_dumpstate(void)
+	uint32_t istackbase = 0;
+	uint32_t istacksize = 0;
 
-	/* Get the limits on the user stack memory */
-
-	ustackbase = (uint32_t)rtcb->adj_stack_ptr;
-	ustacksize = (uint32_t)rtcb->adj_stack_size;
+	/* Get the limits for each type of stack */
 
+	stackbase = (uint32_t)rtcb->adj_stack_ptr;
+	stacksize = (uint32_t)rtcb->adj_stack_size;
 #if CONFIG_ARCH_INTERRUPTSTACK > 3
-	/* Get the limits on the interrupt stack memory */
-
 	istackbase = (uint32_t)&g_intstackbase;
 	istacksize = (CONFIG_ARCH_INTERRUPTSTACK & ~3);
-
-	/* Show interrupt stack info */
-
-	lldbg("sp:     %08x\n", sp);
-	lldbg("IRQ stack:\n");
-	lldbg("  base: %08x\n", istackbase);
-	lldbg("  size: %08x\n", istacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("  used: %08x\n", up_check_intstack());
 #endif
+	bool isIrqAssert = false;
+	bool isSpCorrupt = false;
 
-	/* Does the current stack pointer lie within the interrupt
-	 * stack?
+	/* Check if the assert location is in user thread or IRQ handler.
+	 * If the irq_num is lesser than NVIC_IRQ_USAGEFAULT, then it is
+	 * a fault and not an irq.
 	 */
-
-	if (sp <= istackbase && sp > istackbase - istacksize) {
-		/* Yes.. dump the interrupt stack */
-
-		up_stackdump(sp, istackbase);
+	if (g_irq_nums[2] || (g_irq_nums[1] && g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in nested irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in nested IRQ state!\n");
+	} else if ((g_irq_nums[1] && (g_irq_nums[0] <= NVIC_IRQ_USAGEFAULT)) || (g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in IRQ state!\n");
+	} else {
+		/* Assert in user thread */
+		lldbg("Code asserted in normal thread!\n");
+		if (current_regs) {
+			/* If assert is in user thread, but current_regs is not NULL,
+			 * it means that assert happened due to a fault. So, we want to
+			 * reset the sp to the value just before the fault happened
+			 */
+			sp = current_regs[REG_R13];
+		}
 	}
 
-	/* Extract the user stack pointer if we are in an interrupt handler.
-	 * If we are not in an interrupt handler.  Then sp is the user stack
-	 * pointer (and the above range check should have failed).
-	 */
+	/* Print IRQ handler details if required */
 
-	if (current_regs) {
-		sp = current_regs[REG_R13];
-		lldbg("sp:     %08x\n", sp);
-	}
-
-	lldbg("User stack:\n");
-	lldbg("  base: %08x\n", ustackbase);
-	lldbg("  size: %08x\n", ustacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("  used: %08x\n", up_check_tcbstack(rtcb));
+	if (isIrqAssert) {
+		lldbg("IRQ num: %d\n", g_irq_nums[1]);
+		lldbg("IRQ handler: %08x \n", g_irqvector[g_irq_nums[1]].handler);
+#ifdef CONFIG_DEBUG_IRQ_INFO
+		lldbg("IRQ name: %s \n", g_irqvector[g_irq_nums[1]].irq_name);
 #endif
-
-	/* Dump the user stack if the stack pointer lies within the allocated user
-	 * stack memory.
-	 */
-
-	if (sp <= ustackbase && sp > ustackbase - ustacksize) {
-		up_stackdump(sp, ustackbase);
+		if ((sp <= istackbase) && (sp > (istackbase - istacksize))) {
```

Comment:
You should check the conditional `#if CONFIG_ARCH_INTERRUPTSTACK > 3` here as well.
If it's not enabled, irq will use user stack.

Reply:
Done

Reply:
If we initialize `istackbase` and `istacksize` to zero, then the check for `#if CONFIG_ARCH_INTERRUPTSTACK > 3` is not required. 

However, as Chang mentioned, sp can be of userstack even in case of irq crash. So, let us change our code as below:

```
if (isIrqAssert) {
...
    if sp in irq stack:
    {
        stackbase = istackbase
    }
...
}

if (sp in user stack)
{
...
...
} else {
     isCorruptSp = true;
}

```

Reply:
Updated

Reply:
1. There is no problem in the operation with current. But this will make a SVACE defect. It will say sp checking is always false.
2. Let's assume belows
    1.  `CONFIG_ARCH_INTERRUPTSTACK > 3`
    2.  assert in irq
    In this case, sp should be in irq stack, not in user thread stack. If sp is in user stack, we should print error first. But current code does not.

Reply:
1. Since we are checking sp in istack and stack, atleast one of them will be true and so i think there must be no svace issue. 
2. For this we can change our code as follows:

```
#if CONFIG_ARCH_INTERRUPTSTACK > 3
    if (sp in istack)
        ok.. do the needful
    else
        isSpCorrupted = true
#else
   if (sp in ustack)
        ok.. do the needful
    else
        isSpCorrupted = true
#endif
```